src/omv/modules/py_ml.c

```diff
@@ -63,175 +37,67 @@ static size_t py_ml_tuple_sum(mp_obj_tuple_t *o) {
-
-    int size = (input_width * input_height) - 1; // must be int per countdown loop
+    mp_obj_list_t *input_list = MP_OBJ_TO_PTR(*((mp_obj_t *) arg));
+
+    for (size_t i = 0; i < model->inputs_size; i++) {
+        void *input_buffer = ml_backend_get_input(model, i);
+        size_t input_size = py_ml_tuple_sum(MP_OBJ_TO_PTR(model->input_shape->items[i]));
+        mp_obj_tuple_t *input_shape = MP_OBJ_TO_PTR(model->input_shape->items[i]);
+        mp_obj_t input_arg = input_list->items[i];
+
+        if (mp_obj_is_callable(input_arg)) {
+            // Input is a callable. Call the object and pass the tensor buffer and dtype.
+            mp_obj_t fargs[3] = {
+                mp_obj_new_bytearray_by_ref(input_size, input_buffer),
+                MP_OBJ_FROM_PTR(input_shape),
+                mp_obj_new_int(model->input_dtype)
+            };
+            mp_call_function_n_kw(input_arg, 3, 0, fargs);
+        } else if (MP_OBJ_IS_TYPE(input_arg, &ulab_ndarray_type)) {
+            // Input is an ndarry. The input is converted and copied to the tensor buffer.
+            ndarray_obj_t *input_array = MP_OBJ_TO_PTR(input_arg);
+
+            if (input_array->ndim != input_shape->len) {
+                mp_raise_msg(&mp_type_ValueError,
+                             MP_ERROR_TEXT("Input shape does not match the model input shape"));
+            }
 
-    if (input_channels == 1) {
-        // GRAYSCALE
-        if (model->input_dtype == PY_ML_DTYPE_FLOAT) {
-            // convert u8 -> f32
-            uint8_t *model_input_u8 = (uint8_t *) model_input;
-            float *model_input_f32 = (float *) model_input;
-            for (; size >= 0; size -= 1) {
-                model_input_f32[size] = (model_input_u8[size] * fscale) + fadd;
+            for (size_t i = 0; i < input_array->ndim; i++) {
+                if (input_array->shape[i] != mp_obj_get_int(input_shape->items[i])) {
+                    mp_raise_msg(&mp_type_ValueError,
+                                 MP_ERROR_TEXT("Input shape does not match the model input shape"));
+                }
             }
-        } else {
-            if (shift) {
-                // convert u8 -> s8
-                uint8_t *model_input_8 = (uint8_t *) model_input;
-                #if (__ARM_ARCH > 6)
-                for (; size >= 3; size -= 4) {
-                    *((uint32_t *) (model_input_8 + size - 3)) ^= 0x80808080;
+
+            if (model->input_dtype == 'f') {
+                float *model_input_float = (float *) input_buffer;
+                for (size_t i = 0; i < input_array->len; i++) {
+                    float value = ndarray_get_float_index(input_array->array, input_array->dtype, i);
+                    model_input_float[i] = value;
                 }
-                #endif
-                for (; size >= 0; size -= 1) {
-                    model_input_8[size] ^= PY_ML_GRAYSCALE_MID;
+            } else if (model->input_dtype == 'b') {
```

Comment:
You can merge 'b' and 'B' into one method.

Reply:
I prefer to leave this code as is.